### PR TITLE
[WIP] Connecting to command service to list recordings

### PR DESCRIPTION
**Pull request by ebaron**
This PR is the first attempt at filling in FlightRecorder custom resources with data received from Container JFR. I started with getting a listing of in-memory flight recordings for a JVM. The FlightRecorder reconciler will create a new WebSocket client (if it does not yet exist) and connect to Container JFR's command service. The WebSocket client should then:
1. Query if Container JFR is connected to a JVM already, and disconnect if so.
2. Connect to the JVM pointed to by the reconciler's enqueued FlightRecorder object.
3. Obtain a list of in-memory flight recordings and parse them into RecordingDescriptor structs.
4. Disconnect from the target JVM.

I'm encountering difficulty in step 2 when we try to connect to the target JVM. This seems to run fine for me when testing with a local Container JFR, but when running in OpenShift, I get repeated `close 1000 (normal)` control messages upon reading. Perhaps something is timing out here, because it takes a while to attempt to connect. Eventually, I seem to get a response for the inital connect request that has an exception like this: `WrappedConnectionException: Failed to retrieve RMIServer stub: javax.naming.ServiceUnavailableException [Root exception is java.rmi.ConnectException: Connection refused to host: 172.30.30.231; nested exception is: \n\tjava.net.ConnectException: Connection timed out (Connection timed out)]`. I checked to make sure the IP address and port correspond to the running Container JFR's service that exposes port 9091 for JMX.

I've attached the relevant logs for runs of the operator and Container JFR. Any help getting to the bottom of this issue would be much appreciated, I've been stumped for some time.

[container-jfr-log.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/3806647/TMP-new-container-jfr-log.txt)
[operator-log.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/3806648/TMP-new-operator-log.txt)

### CODE REVIEW

**Comment by andrewazores** [comment]
What kind of OpenShift environment are you using for testing? I'll first make sure I can also reproduce the issue in the same sort of environment.

**Comment by ebaron** [comment]
> What kind of OpenShift environment are you using for testing? I'll first make sure I can also reproduce the issue in the same sort of environment.

I'm testing with OpenShift 3.11 using Minishift, I still lack the resources to run OpenShift 4.x locally. With the Operator SDK upgrade, I needed to hack the CRD files to allow them to deploy properly. OpenShift 3.11 doesn't like there to be a `type` attribute under `openAPIV3Schema`. I just modified the Makefile as follows for now:
```make
.PHONY: openapi
openapi:
    operator-sdk generate openapi
    sed -i 's/^      type: object//g' deploy/crds/rhjmc.redhat.com_*crd.yaml
```

**Comment by andrewazores** [comment]
> > What kind of OpenShift environment are you using for testing? I'll first make sure I can also reproduce the issue in the same sort of environment.
> 
> I'm testing with OpenShift 3.11 using Minishift, I still lack the resources to run OpenShift 4.x locally. With the Operator SDK upgrade, I needed to hack the CRD files to allow them to deploy properly. OpenShift 3.11 doesn't like there to be a `type` attribute under `openAPIV3Schema`. I just modified the Makefile as follows for now:
> 
> ```makefile
> .PHONY: openapi
> openapi:
>     operator-sdk generate openapi
>     sed -i 's/^      type: object//g' deploy/crds/rhjmc.redhat.com_*crd.yaml
> ```

Cool, I'll test it out right now.

File: pkg/controller/flightrecorder/flightrecorder_controller.go [hunk 1]
```diff
@@ -73,9 +77,30 @@ func (r *ReconcileFlightRecorder) Reconcile(request reconcile.Request) (reconcil
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	reqLogger.Info("Reconciling FlightRecorder")
 
+	commandSvc := &corev1.Service{}
+	commandSvcName := "containerjfr-command"                                                                       // TODO make const or get from ContainerJFR
```
[the remark this answers is not in the document]
Comment (ebaron) [reply pkg/controller/flightrecorder/flightrecorder_controller.go]: Ah sorry, I see what you mean. I'll clarify the TODO for now.

**Comment by andrewazores** [comment]
This is what I think I see happening in my test scenario - let me know if any assumptions are incorrect.

1) `make deploy` creates the CRDs etc. and deploys the Operator
2) The Operator comes online and examines the state of the cluster
3) The Operator creates a ContainerJFR instance due to the CR created in the last step of the `make deploy`
4) This ContainerJFR's associated Service is then noticed by the Operator on its next loop, and is detected as being JFR-aware since it advertises the expected port
5) The Operator creates a FlightRecorder CR for the ContainerJFR Service
6) While reconciling the FlightRecorder, the Operator attempts to open a WebSocket connection to ContainerJFR, asking it to connect to the target service (ie itself)
7) ???

At step 7, ContainerJFR appears to never respond and the Operator is left wondering what's happening, potentially with its WebSocket connection to the ContainerJFR instance timing out and retrying. Does this match up with your series of events?

If so, I think we're seeing an old bug (which I thought was resolved) where asking ContainerJFR to connect to itself via its Cluster IP causes it to hang. I guess the fix as implemented only applied to OpenShift 4, since I tested it with CRC and not Minishift?

I'm going to modify the flightrecorder_controller so that it avoids connecting ContainerJFR to itself and see if it's able to work as expected for the sample_app deployment. I'll also test it as-is in CRC and see what happens.

If the problem is what I think it is then we need to figure out what the root cause of the connection issue is with OpenShift 3, and either address that directly, or change the advertised connection URL for Services when running in OpenShift 3 (ie to something other than Cluster IP). I'll also see if I can find a way to set a timeout on the connection attempt from within ContainerJFR, so that it can bail rather than hanging if any scenario like this comes up with any target JVM.

**Comment by andrewazores** [comment]
I added a cheap hack to the service_controller to skip creation of the FlightRecorder if the requested service has label name=containerjfr, and this allowed it to progress to detecting and reconciling for the jmx-listener sample app. I also need to add a hack here because that sample app actually listens on port 9093 (I did this so that I could run it as a local JVM process alongside a local container-jfr process for a desktop demo), so I modified the isJFRAware check to also check port 9093 and modified the flightrecorder_controller to request container-jfr to connect to the target service on that port.

[containerjfr.log](https://github.com/rh-jmc-team/container-jfr-operator/files/3810816/containerjfr.log)

[operator.log](https://github.com/rh-jmc-team/container-jfr-operator/files/3810817/operator.log)

**Comment by andrewazores** [comment]
Testing this PR as-is with CRC shows it working just fine, with ContainerJFR detecting and subsequently connecting to itself. So, whatever hanging issue seems to exist is somehow specific to the networking setup in OpenShift 3, since it doesn't appear in OpenShift 4, or locally run container-jfr, or container-jfr run with a local Docker daemon, etc.

**Comment by ebaron** [comment]
Your assumptions above are correct. I tried to do a similar test where Container JFR itself is excluded by adding the following to the FlightRecorder Reconcile method:
```go
	targetRef := instance.Status.Target
	if targetRef.Name == "containerjfr" {
		return reconcile.Result{}, nil // XXX
	}
```

I used `quay.io/rh-jmc-team/hot-methods:2` as a test application. I still saw the same behaviour on Minishift that I did when trying to connect to ContainerJFR's JVM :disappointed: 

[hotmethods-operator.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/3811092/hotmethods-operator.txt)
[hotmethods-containerjfr.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/3811093/hotmethods-containerjfr.txt)

**Comment by andrewazores** [comment]
Huh... I'll try to replicate that too then.

If you set up that same test scenario but with a previous version of the Operator which doesn't establish the WebSocket connection, are you able to connect to the hot-methods application via the -web UI?

You can also use something like https://github.com/vi/websocat to connect to the exposed -command route, which might be useful to diagnose if the problem is with the operator or with container-jfr.

**Comment by andrewazores** [comment]
I'm running the same setup with the `targetRef` check and the hot-methods sample app, and I also see very similar logs to what you shared.

The first type of connection failure, the `{"command":"connect","args":["172.30.185.123:9091"]},"error":"websocket: close 1000 (normal)" ones, seem to go away after some time in both of our logs, and we get `"request":"myproject/hot-methods","error":"server failed to execute \"connect\": WrappedConnectionException: Failed to retrieve RMIServer stub: javax.naming.ServiceUnavailableException [Root exception is java.rmi.ConnectException: Connection refused to host: 172.30.185.123; nested exception is: \n\tjava.net.ConnectException: Connection refused (Connection refused)] (code -2)"` instead. That looks like a different problem - it looks like the operator is able to connect to container-jfr fine at this point, but container-jfr is unable to open a connection to the target.

I tried using websocat to communicate with container-jfr directly on its exposed route, and the connection established normally. `scan-targets` returned the expected results, including `{"connectUrl":"hot-methods.myproject.svc.cluster.local","alias":"hot-methods.myproject.svc.cluster.local","port":9091}`. Sending a `{"command":"connect","args":["hot-methods.myproject.svc.cluster.local:9091"]}` returned an error response with the same "connection refused" that shows up in the operator logs. Hmm...

**Comment by andrewazores** [comment]
Here's another "interesting" failure mode I just hit:

```
{"level":"info","ts":1573052897.7911675,"logger":"controller_flightrecorder","msg":"Reconciling FlightRecorder","Request.Namespace":"myproject","Request.Name":"jmx-listener"}
E1106 15:08:17.791418       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/andrew/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:76
/home/andrew/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:65
/home/andrew/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:51
/usr/lib/golang/src/runtime/panic.go:522
/usr/lib/golang/src/runtime/panic.go:82
/usr/lib/golang/src/runtime/signal_unix.go:390
container-jfr-operator/pkg/controller/flightrecorder/flightrecorder_controller.go:105
```

This is on the `targetRef := instance.Status.Target` line, so I'm not sure if it was the Status or the Target that was nil.

**Comment by andrewazores** [comment]
I think there's just something up with the `hot-methods:2` image. The Operator can't connect to it properly, but neither can I when interacting with the container-jfr instance using websocat. If I use `quay.io/andrewazores/container-jmx-docker-listener:1` as my sample app (this one has the JMX port put back to 9091) then it again works just fine, both using websocat and for the Operator. Could you try using that tag for your sample app too?

**Comment by ebaron** [comment]
> Here's another "interesting" failure mode I just hit:
> 
> ```
> {"level":"info","ts":1573052897.7911675,"logger":"controller_flightrecorder","msg":"Reconciling FlightRecorder","Request.Namespace":"myproject","Request.Name":"jmx-listener"}
> E1106 15:08:17.791418       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
> /home/andrew/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:76
> /home/andrew/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:65
> /home/andrew/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:51
> /usr/lib/golang/src/runtime/panic.go:522
> /usr/lib/golang/src/runtime/panic.go:82
> /usr/lib/golang/src/runtime/signal_unix.go:390
> container-jfr-operator/pkg/controller/flightrecorder/flightrecorder_controller.go:105
> ```
> 
> This is on the `targetRef := instance.Status.Target` line, so I assume the Status was invalid somehow. Unless the `instance` became invalid after the `Get` call?

Hmmm, that's interesting. Status isn't a pointer so that shouldn't be a problem. I think the only thing that would make sense is if instance was a nil-pointer. We initialize this, so it would have to be client.Get changing it to nil somehow.

**Comment by ebaron** [comment]
> I think there's just something up with the `hot-methods:2` image. The Operator can't connect to it properly, but neither can I when interacting with the container-jfr instance using websocat. If I use `quay.io/andrewazores/container-jmx-docker-listener:1` as my sample app (this one has the JMX port put back to 9091) then it again works just fine, both using websocat and for the Operator. Could you try using that tag for your sample app too?

I created the test app using `oc new-app quay.io/andrewazores/container-jmx-docker-listener:1 --name=jmx-listener`. (Just as an aside, I omitted the `--name` to begin with and got an error from Container JFR: `WrappedConnectionException: Unknown host: jmx-listener; nested exception is: \n\tjava.net.UnknownHostException: jmx-listener`)

I initially ran into the same problem, but after timing out once, it succeeded. Seems like some kind of timing issue.

[jmx-listener-containerjfr.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/3816408/jmx-listener-containerjfr.txt)
[jmx-listener-operator.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/3816409/jmx-listener-operator.txt)

**Comment by andrewazores** [comment]
> > I think there's just something up with the `hot-methods:2` image. The Operator can't connect to it properly, but neither can I when interacting with the container-jfr instance using websocat. If I use `quay.io/andrewazores/container-jmx-docker-listener:1` as my sample app (this one has the JMX port put back to 9091) then it again works just fine, both using websocat and for the Operator. Could you try using that tag for your sample app too?
> 
> I created the test app using `oc new-app quay.io/andrewazores/container-jmx-docker-listener:1 --name=jmx-listener`. (Just as an aside, I omitted the `--name` to begin with and got an error from Container JFR: `WrappedConnectionException: Unknown host: jmx-listener; nested exception is: \n\tjava.net.UnknownHostException: jmx-listener`)

Ah, yea, I forgot to mention that part, sorry. That's a holdover from that demo also being run in different Docker and Podman setups.

> 
> I initially ran into the same problem, but after timing out once, it succeeded. Seems like some kind of timing issue.
> 
> [jmx-listener-containerjfr.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/3816408/jmx-listener-containerjfr.txt)
> [jmx-listener-operator.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/3816409/jmx-listener-operator.txt)

Nice, I guess? Timing issue isn't fun, but at least it's in an event loop and can simply be requeued if that kind of unexpected closure occurs. At least this proves some degree of robustness exists! Haha

**Comment by ebaron** [comment]
Closed in favour of #44

**Pull request closed without merging**